torch/csrc/autograd/record_function.h

```diff
@@ -82,20 +128,27 @@ struct TORCH_API RecordFunction {
   Function* fn_ = nullptr;
   StringView name_;
   int64_t sequence_nr_ = -1;
-
+  std::vector<c10::IValue> inputs_;
   RecordFunction* parent_ = nullptr;
 
-  GetPackedInputsCallback inputs_cb_ = nullptr;
-  mutable bool inputs_initialized_ = false;
-  // initialized lazily by inputs_cb_
-  mutable std::vector<c10::IValue> inputs_;
+  RecordFunctionState state_;
+  bool initialized_ = false;
 };
 
+#define RECORD_FUNCTION(fn, ...) \
+  torch::autograd::profiler::RecordFunction guard; \
+  if (torch::autograd::profiler::recordFunctionState() != \
+      torch::autograd::profiler::RecordFunctionState::NO_CALLBACKS) { \
+    guard.before(fn, __VA_ARGS__); \
```

Comment:
This is constructing IValues whenever _any_ profiling is enabled. IValue construction is not free. It requires boxing primitives and refcounting on tensors.  It is really careful to figure out whether we should be recording inputs, and then it doesn't use that information for the most expensive part of recording those inputs.